var/spack/repos/builtin/packages/elpa/package.py

```diff
@@ -62,21 +65,48 @@ def libs(self):
 
     build_directory = 'spack-build'
 
-    def setup_environment(self, spack_env, run_env):
+    def configure_args(self):
         # TODO: set optimum flags for platform+compiler combo, see
         # https://github.com/hfp/xconfigure/tree/master/elpa
-        spec = self.spec
+        # also see:
+        # https://src.fedoraproject.org/cgit/rpms/elpa.git/
+        # https://packages.qa.debian.org/e/elpa.html
 
-        spack_env.set('CC', spec['mpi'].mpicc)
-        spack_env.set('FC', spec['mpi'].mpifc)
-        spack_env.set('CXX', spec['mpi'].mpicxx)
+        options = []
 
-        spack_env.append_flags('LDFLAGS', spec['lapack'].libs.search_flags)
-        spack_env.append_flags('LIBS', spec['lapack'].libs.link_flags)
-        spack_env.set('SCALAPACK_LDFLAGS', spec['scalapack'].libs.joined())
+        spec = self.spec
+        env['CC'] = spec['mpi'].mpicc
+        env['CXX'] = spec['mpi'].mpicxx
+        env['F77'] = spec['mpi'].mpif77
+        env['FC'] = spec['mpi'].mpifc
+
+        if spec.satisfies('%intel'):
+            options.append('CPP=gcc -E')
 
-    def configure_args(self):
-        options = []
         if '+openmp' in self.spec:
             options.append("--enable-openmp")
+
+        if '+optflags' in self.spec:
```

Comment:
I have these optimization flags in my `compilers.yaml` file indeed. I included here just to follow the spirit this [PR](https://github.com/spack/spack/pull/5306), and to make the optimization flags more complete when it is requested by `+optflags`.

However, since I'm just getting started with Spack, so very likely I misunderstood certain things. I thought Spack's foremost use case is for HPC environment. Then I'd think turning on optimizations and using highest instruction sets for the hardware will be very important; say, a 20% loss of efficiency over thousands or tens of thousands cores will be quite substantial. Globally using `-O3` is not always OK for all packages, so it would seem beneficial to have some "highest" possible optimization flags built in to each package.

Reply:
> Globally using -O3 is not always OK for all packages, so it would seem beneficial to have some "highest" possible optimization flags built in to each package.

Imho, that doesn't scale with the architectures. For instance, why should I assume to have AVX available if `+opt`? 

> Globally using -O3 is not always OK for all packages, so it would seem beneficial to have some "highest" possible optimization flags built in to each package.

Side note: this gives me the impression that you think `-O3` is a deterministic set of rules. `-O3` is just a convenience flag that maps to different optimizations for different compilers and different versions of the same compiler. This I think reinforces my conviction that optimization flags should be separated from packages and reside in `compilers.yaml` or `packages.yaml`.

Pinging @lee218llnl and @adamjstewart so that they can also give their opinion on this (which may be different from mine).

Reply:
I agree that we can't assume the highest possible optimizations and instruction set. For one, -O3 isn't always beneficial over -O2, at least for the Intel compilers. The -O3 optimizations generally lead to a larger executable and in some cases can run slower than -O2. Also, @alalazo is correct about things like AVX. At LLNL we have several different generations of architectures sharing the same networked file system and thus we prefer sticking to the lowest common denominator. Plus, we don't want a new Spack user to install a package on their old system and wonder why they are getting illegal instructions.

Regarding rpaths, yes, I think the assumption is that you would rebuild everything on the new machine if the paths are not going to be the same. Perhaps one could conceive of a way to chrpath a bunch of executables and libraries, but that can get messy. Anyway, don't we have functionality in Spack to create a recipe/specification with a bunch of packages? This isn't something I have had to worry about so I don't remember the details or if I'm making this up. @tgamblin @becker33?

Reply:
> This I think reinforces my conviction that optimization flags should be separated from packages and reside in `compilers.yaml` or `packages.yaml`.

:+1:

Flags that are needed to get something to install at all should go in `package.py` files. Flags that are convenient and that you personally prefer to use should go in `compilers.yaml`.

Reply:
@alalazo @lee218llnl @adamjstewart Thanks for all your comments! I wanted to clarify that I was not advocating universal `-O3` over `-O2` and similar. I understand that `-O3` might include a different set of optimizations for different compilers/versions, might not always outperform `-O2` for certain packages, and might even break others. This kind of knowledge is something that packagers much more likely to possess than end users, and it's exactly why I think packagers should include the highest possible optimization levels that they know work for their packages (on most/typical machines), or at least include an option to enable so. Otherwise, end users have to decide what's the best flags for packageA (`-O3 ...`), for packageB (`-O2 ...`), etc. It's then also not sufficient to just modify `compilers.yaml` but `packages.yaml` as well.

I think the approach of including the lowest denominator in the distributions while letting users customize their compiling preference fits best with Linux distributions/FreeBSD that target use cases where bottleneck is typically not computation, but seems at odds with Spack positioning as the package manager for **HPC** that end users can just `spack install` away. 

For example, I used Spack to compile a particular package **after** setting some typical optimization flags in `compiler.yaml`. It turns out the executable is slower than what my colleague manually compiled by about 15% (compilers and libraries used are the same in both cases), which made me have to go through all the dependency packages and figure out what might be causing the difference. Using a code 15% slower on hundreds or thousands of nodes for months for a project seems too wasteful.

Regarding different machines sharing the same networked file system, is there a way to tell Spack to build to different locations? Forgoing `AVX` instruction sets on newer machines because they share the same file system with some old machines with, say, only `SSE` is probably even worse than not using the best set of optimization flags. And for the avx flags for `elpa`, what's the best way to probe CPU capability with Spack?

Reply:
Out of curiosity: Did you find out where the performance drop originated?

Reply:
For example, with this package, `--with-avx-optimization` will make a difference, but the performance difference I was referring to was another package that depends on a few other math libraries, all of which have their own instructions on optimizations.